File: src/main/java/net/hectus/neobb/turn/default_game/warp/WarpTurn.java
package net.hectus.neobb.turn.default_game.warp;

import com.marcpg.libpg.storing.Cord;
import com.marcpg.libpg.storing.CordMinecraftAdapter;
import com.marcpg.libpg.util.Randomizer;
import net.hectus.neobb.player.NeoPlayer;
import net.hectus.neobb.structure.PlacedStructure;
import net.hectus.neobb.structure.Structure;
import net.hectus.neobb.structure.StructureManager;
import net.hectus.neobb.turn.default_game.attributes.clazz.Clazz;
import net.hectus.neobb.turn.default_game.attributes.function.WarpFunction;
import net.hectus.neobb.turn.default_game.structure.StructureTurn;
import net.hectus.neobb.util.Configuration;
import net.hectus.neobb.util.Modifiers;
import org.bukkit.Location;
import org.bukkit.World;

import java.util.List;

public abstract class WarpTurn extends StructureTurn implements WarpFunction {
    public final Location center;
    public final String name;

    public WarpTurn(World world, String name) {
        super(null);
        this.cord = Cord.ofList(Configuration.CONFIG.getIntegerList("warps." + name));
        this.name = name;
        this.center = CordMinecraftAdapter.toLocation(cord, world).add(5, 0, 5);
    }

    public WarpTurn(PlacedStructure data, World world, NeoPlayer player, String name) {
        super(data, player);
        this.cord = Cord.ofList(Configuration.CONFIG.getIntegerList("warps." + name));
        this.name = name;
        this.center = CordMinecraftAdapter.toLocation(cord, world).add(5, 0, 5);
    }

    public abstract int chance();
    public abstract List<Class<? extends Clazz>> allows();

    public boolean canBePlayed() {
        return true;
    }

    public enum Temperature { COLD, NORMAL, HOT }
    public Temperature temperature() {
        return switch (allows().getFirst().getSimpleName()) {
            case "ColdClazz" -> Temperature.COLD;
            case "HotClazz" -> Temperature.HOT;
            default -> Temperature.NORMAL;
        };
    }

    public Cord highCorner() {
        return cord.add(new Cord(9, Configuration.MAX_ARENA_HEIGHT, 9));
    }

    @Override
    public boolean requiresUsageGuide() {
        return true;
    }

    @Override
    public int maxAmount() {
        return 1;
    }

    @Override
    public void apply() {
        if (canBePlayed() && (Randomizer.boolByChance(chance()) || player.hasModifier(Modifiers.P_DEFAULT_100P_WARP))) {
            player.removeModifier(Modifiers.P_DEFAULT_100P_WARP);
            player.game.players().forEach(p -> p.teleport(p.cord().subtract(player.game.warp().cord()).add(cord)));
            player.game.warp(this);
        }
    }

    @Override
    public Structure referenceStructure() {
        return StructureManager.structure(name + "-warp");
    }
}

File: src/main/kotlin/net/hectus/neobb/turn/default_game/warp/WarpTurn.kt
package net.hectus.neobb.turn.default_game.warp

import com.marcpg.libpg.storing.Cord
import com.marcpg.libpg.util.Randomizer
import net.hectus.neobb.player.NeoPlayer
import net.hectus.neobb.structure.PlacedStructure
import net.hectus.neobb.turn.default_game.attribute.clazz.Clazz
import net.hectus.neobb.turn.default_game.attribute.function.WarpFunction
import net.hectus.neobb.turn.default_game.structure.StructureTurn
import net.hectus.neobb.util.Configuration
import net.hectus.neobb.util.Modifiers
import kotlin.reflect.KClass

abstract class WarpTurn(data: PlacedStructure?, cord: Cord, val name: String, player: NeoPlayer?) : StructureTurn(data, cord, player), WarpFunction {
    enum class Temperature { COLD, NORMAL, HOT }

    override val requiresUsageGuide: Boolean = true
    override val maxAmount: Int = 1

    abstract val chance: Double
    abstract val allows: List<KClass<out Clazz>>
    abstract val temperature: Temperature

    val lowCorner: Cord = Cord.ofList(Configuration.CONFIG.getIntegerList("warps.$name"))
    val center: Cord = lowCorner.add(Cord(4.5, 0.0, 4.5))
    val highCorner: Cord = lowCorner.add(Cord(9.0, Configuration.MAX_ARENA_HEIGHT.toDouble(), 9.0))

    open fun canBePlayed(): Boolean = true

    override fun apply() {
        if (canBePlayed() && (Randomizer.boolByChance(chance) || player!!.hasModifier(Modifiers.Player.Default.ALWAYS_WARP))) {
            player!!.removeModifier(Modifiers.Player.Default.ALWAYS_WARP)
            player.game.players.forEach { it.teleport(it.cord().subtract(player.game.warp.lowCorner).add(lowCorner)) }
            player.game.warp(this)
        }
    }
}
